### PR TITLE
Ap to kp

### PR DESCRIPTION
**Pull request by aburrell**
Added routine to convert from ap to kp.

### CODE REVIEW

**Comment by rstoneback** [comment]
Thanks for the code! Headed to the doctor but I’ll take a look after.

Cheers,
Russell

On Jul 1, 2019, at 2:21 PM, Angeline Burrell <notifications@github.com<mailto:notifications@github.com>> wrote:


@aburrell<https://github.com/aburrell> requested your review on: #245<https://github.com/rstoneback/pysat/pull/245> Ap to kp.

File: pysat/instruments/methods/sw.py [hunk 1]
```diff
@@ -511,3 +511,74 @@ def calc_daily_Ap(ap_inst, ap_name='3hr_ap', daily_name='Ap',
...
+    ap_keys = sorted(list(ap_to_kp.keys()))
+
+    # If the ap falls between two Kp indices, assign it to the lower Kp value
+    def round_ap(ap_in, fill_val=fill_val):
+        """ Round an ap value to the nearest Kp value
+        """
+        if not np.isfinite(ap_in):
+            return fill_val
+        
+        i = 0
```
Comment (rstoneback) [inline pysat/instruments/methods/sw.py]: Starting at line 555 (here) down to the return aren't tested with the unit tests.

Comment (aburrell) [reply pysat/instruments/methods/sw.py]: Added three more unit tests for these cases.

File: pysat/tests/test_sw.py [hunk 2]
```diff
@@ -16,13 +16,21 @@ def setup(self):
...
+                                                            fill_value=np.nan),
+                                            'ap_inf': np.full(shape=12, \
+                                                            fill_value=np.inf)},
                                            index=[pysat.datetime(2009, 1, 1)
                                                   + pds.DateOffset(hours=3*i)
                                                   for i in range(12)])
         self.testInst.meta = pysat.Meta()
         self.testInst.meta.__setitem__('Kp', {self.testInst.meta.fill_label:
                                               np.nan})
+        self.testInst.meta.__setitem__('ap_nan', {self.testInst.meta.fill_label:
```
Comment (rstoneback) [inline pysat/tests/test_sw.py]: These could be simplified,

        self.testInst.meta['Kp'] = {self.testInst.meta.fill_label:
                                              np.nan}
        self.testInst.meta['ap_nan'] = {self.testInst.meta.fill_label:
                                                  np.nan}
        self.testInst.meta['ap_inv'] = {self.testInst.meta.fill_label:
                                                  np.inf}

File: pysat/instruments/methods/sw.py [hunk 3]
```diff
@@ -511,3 +511,74 @@ def calc_daily_Ap(ap_inst, ap_name='3hr_ap', daily_name='Ap',
     ap_inst.meta.__setitem__(daily_name, meta_dict)
```
Comment (rstoneback) [inline pysat/instruments/methods/sw.py]: This could also be modified, though technically not in pull.
ap_inst.meta[daily_name] =  meta_dict

**Comment by rstoneback** [comment]
The tests don't pass on my local machine for some reason.


======================================================================
ERROR: Test combine_f107 when all input is provided with '' and '45day'
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/russellstoneback/git/pysat/pysat/tests/test_sw.py", line 498, in test_combine_f107_all
    **self.combineTimes)
  File "/Users/russellstoneback/git/pysat/pysat/instruments/methods/sw.py", line 381, in combine_f107
    freq = pysat.utils.time.calc_freq(f107_times)
  File "/Users/russellstoneback/git/pysat/pysat/utils/time.py", line 100, in calc_freq
    raise ValueError("insufficient data to calculate frequency")
ValueError: insufficient data to calculate frequency

======================================================================
ERROR: Test combine_f107 with times provided through datasets
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/russellstoneback/git/pysat/pysat/tests/test_sw.py", line 484, in test_combine_f107_inst_time
    self.combineInst['forecast'])
  File "/Users/russellstoneback/git/pysat/pysat/instruments/methods/sw.py", line 289, in combine_f107
    " starting and ending times")
ValueError: must either load in Instrument objects or provide starting and ending times

======================================================================
ERROR: Test combine_kp when times are provided through the instruments
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/russellstoneback/git/pysat/pysat/tests/test_sw.py", line 276, in test_combine_kp_inst_time
    combo_in['stop'] = combo_in['forecast_inst'].index[-1]
  File "/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/pandas/core/indexes/base.py", line 3957, in __getitem__
    return getitem(key)
IndexError: index -1 is out of bounds for axis 0 with size 0

======================================================================
ERROR: Test combine_kp when standard data is not provided
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/russellstoneback/git/pysat/pysat/tests/test_sw.py", line 349, in test_combine_kp_no_standard
    kp_inst = sw_meth.combine_kp(**combo_in)
  File "/Users/russellstoneback/git/pysat/pysat/instruments/methods/sw.py", line 196, in combine_kp
    freq = pysat.utils.time.calc_freq(kp_times)
  File "/Users/russellstoneback/git/pysat/pysat/utils/time.py", line 100, in calc_freq
    raise ValueError("insufficient data to calculate frequency")
ValueError: insufficient data to calculate frequency

======================================================================
FAIL: Test combine_kp when all input is provided
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/russellstoneback/Library/Enthought/Canopy/edm/envs/Python_2_again/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/russellstoneback/git/pysat/pysat/tests/test_sw.py", line 304, in test_combine_kp_all
    assert (kp_inst['Kp'] != self.combine['fill_val']).all()
AssertionError

**Comment by aburrell** [comment]
But they're passing globally, and were not changed by this pull request.

**Comment by rstoneback** [comment]
Just a bit confused. @jklenzing do you have an idea why this might happen? Could this be related to some of your testing changes?

**Comment by jklenzing** [comment]
@rstoneback, potentially.  What command are you using to run the unit tests?  ie, are you running a subset, or the full thing? Will try to duplicate here.

**Comment by rstoneback** [comment]
nosetests -vs pysat.tests.test_sw

**Pull request merged**